### PR TITLE
Add broadcast function to the API

### PR DESCRIPTION
**Pull request by shoyer**
This is a renaming and update of the existing `xray.broadcast_arrays` function,
which now works properly in the light of #648.

xref #649
cc @rabernat 

Examples
--------

Broadcast two data arrays against one another to fill out their dimensions:

    >>> a = xray.DataArray([1, 2, 3], dims='x')
    >>> b = xray.DataArray([5, 6], dims='y')
    >>> a
    <xray.DataArray (x: 3)>
    array([1, 2, 3])
    Coordinates:
      * x        (x) int64 0 1 2
    >>> b
    <xray.DataArray (y: 2)>
    array([5, 6])
    Coordinates:
      * y        (y) int64 0 1
    >>> a2, b2 = xray.broadcast(a, b)
    >>> a2
    <xray.DataArray (x: 3, y: 2)>
    array([[1, 1],
           [2, 2],
           [3, 3]])
    Coordinates:
      * x        (x) int64 0 1 2
      * y        (y) int64 0 1
    >>> b2
    <xray.DataArray (x: 3, y: 2)>
    array([[5, 6],
           [5, 6],
           [5, 6]])
    Coordinates:
      * y        (y) int64 0 1
      * x        (x) int64 0 1 2

Fill out the dimensions of all data variables in a dataset:

    >>> ds = xray.Dataset({'a': a, 'b': b})
    >>> ds2, = xray.broadcast(ds)  # use tuple unpacking to extract one dataset
    >>> ds2
    <xray.Dataset>
    Dimensions:  (x: 3, y: 2)
    Coordinates:
      * x        (x) int64 0 1 2
      * y        (y) int64 0 1
    Data variables:
        a        (x, y) int64 1 1 2 2 3 3
        b        (x, y) int64 5 6 5 6 5 6

### CODE REVIEW

**Comment by shoyer** [comment]
Any comments? This needs a rebase, but is otherwise ready for review

File: xray/test/test_dataarray.py [hunk 1]
```diff
@@ -1276,15 +1276,15 @@ def test_broadcast_arrays(self):
 
         x = DataArray(np.random.randn(2, 3), dims=['a', 'b'])
         y = DataArray(np.random.randn(3, 2), dims=['b', 'a'])
-        x2, y2 = broadcast_arrays(x, y)
+        x2, y2 = broadcast(x, y)
```
Comment (rabernat) [inline xray/test/test_dataarray.py]: Can there be a test for broadcasting coordinates, not just data variables? That is something I use frequently, and I currently have to hack it (as in #649).

File: xray/test/test_dataset.py [hunk 2]
```diff
@@ -937,6 +937,30 @@ def test_align(self):
...
+                            'baz': (('x', 'y'), [[2, 3]])},
+                            {'c': ('x', [4])})
+        actual, = broadcast(ds)
+        self.assertDatasetIdentical(expected, actual)
+
+        ds_x = Dataset({'foo': ('x', [1])})
+        ds_y = Dataset({'bar': ('y', [2, 3])})
+        expected_x = Dataset({'foo': (('x', 'y'), [[1, 1]])})
+        expected_y = Dataset({'bar': (('x', 'y'), [[2, 3]])})
+        actual_x, actual_y = broadcast(ds_x, ds_y)
```
Comment (rabernat) [inline xray/test/test_dataset.py]: Never mind the previous comment. It looks like you test coordinate broadcasting here.

File: xray/test/test_dataarray.py (hunk 1 above)
Comment (shoyer) [inline xray/test/test_dataarray.py]: I'll add some more explicit tests.

File: xray/core/alignment.py [hunk 3]
```diff
@@ -276,14 +325,47 @@ def broadcast_arrays(*args):
...
+
+    def _broadcast_array(array):
+        data = array.variable.expand_dims(dims_map)
+        coords = OrderedDict(array.coords)
+        coords.update(common_coords)
+        dims = tuple(common_coords)
+        name = array.name
+        attrs = array.attrs
+        encoding = array.encoding
+        return DataArray(data, coords, dims, name, attrs, encoding)
```
Comment (jhamman) [inline xray/core/alignment.py]: unpacking all the attrs here seems unnecessarily verbose, not wrong but just more verbose than I'm used to seeing.

**Comment by jhamman** [comment]
LGTM.

File: xray/core/alignment.py (hunk 3 above)
Comment (shoyer) [inline xray/core/alignment.py]: fixed

File: xray/test/test_dataarray.py (hunk 1 above)
Comment (shoyer) [inline xray/test/test_dataarray.py]: done

**Comment by shoyer** [comment]
thanks for the feedback, guys!

**Pull request merged**